lib/Backend/arm64/LowerMD.cpp

```diff
@@ -3294,40 +3294,29 @@ LowererMD::GenerateFastAdd(IR::Instr * instrAdd)
         this->GenerateSmIntPairTest(instrAdd, opndSrc1, opndSrc2, labelHelper);
     }
 
-    if (opndSrc1->IsAddrOpnd())
-    {
-        // If opnd1 is a constant, just swap them.
-        IR::Opnd *opndTmp = opndSrc1;
-        opndSrc1 = opndSrc2;
-        opndSrc2 = opndTmp;
-    }
-
     //
     // For 32 bit arithmetic we copy them and set the size of operands to be 32 bits. This is
     // relevant only on ARM64.
     //
 
     opndSrc1 = opndSrc1->UseWithNewType(TyInt32, this->m_func);
-    
+    opndSrc2 = opndSrc2->UseWithNewType(TyInt32, this->m_func);
+
     // s1 = MOV src1
 
     opndReg = IR::RegOpnd::New(TyInt32, this->m_func);
-    instr = IR::Instr::New(Js::OpCode::MOV, opndReg, opndSrc1, this->m_func);
-    instrAdd->InsertBefore(instr);
-    
-    if (opndSrc2->IsAddrOpnd())
-    {
-        // truncate to untag
-        int value = ::Math::PointerCastToIntegralTruncate<int>(opndSrc2->AsAddrOpnd()->m_address);
-        opndSrc2 = IR::IntConstOpnd::New(value, TyInt32, this->m_func);
-        instr = IR::Instr::New(Js::OpCode::ADDS, opndReg, opndReg, opndSrc2, this->m_func);
-    }
-    else
+    Lowerer::InsertMove(opndReg, opndSrc1, instrAdd);
+
+    if (opndSrc2->IsImmediateOpnd())
     {
-        instr = IR::Instr::New(Js::OpCode::ADDS, opndReg, opndReg, opndSrc2->UseWithNewType(TyInt32, this->m_func), this->m_func);
+        IR::Opnd* tmp = IR::RegOpnd::New(TyInt32, this->m_func);
+        instr = IR::Instr::New(Js::OpCode::LDIMM, tmp, opndSrc2, this->m_func);
+        instrAdd->InsertBefore(instr);
+        opndSrc2 = tmp;
     }
```

Comment:
Seems odd to have the immediate logic explicitly here when the Legalizer should detect immediates that can't be encoded and automatically do this. Also here you're doing it unconditionally. You could at least check to see if you need to first.

Reply:
I found cases in our unit tests were SUBS was reaching the encoder with an src2 IntConstOpnd that didn't get hoisted by the legalizer and couldn't be encoded. I forgot the legalizer was supposed to handle that so I changed it in GenerateFastSub, and since I had a different issue to fix in GenerateFastAdd I thought I'd move towards consolidating the very similar logic between the two.

You're right, fixing the issue in the legalizer is definitely the way to go. That change wasn't supposed to be part of though, I got my branches mixed up, so I'll update the other PR to fix that issue in the legalizer and remake this PR without the branch confusion.